stream: abort signal in readable streams

benjamingr opened the pull request:
Mostly opened for discussion about semantics.

Are these semantics correct? Do they make sense?

This is WIP: I want to test this thoroughly before considering merging + docs need to be updated. 

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

benjamingr commented:
Idea from Matteo:

 - Expose an `.abortSignal` property on `stream.Readable` with:
   - A getter that returns the signal if present or `null` otherwise.
   - A setter that sets the abort signal on the stream.

Question: what if someone sets it multiple times does it unsubscribe from previous calls?

Maybe a static helper for setting an abort signal?

ronag commented:
I think it’s better without setter.

benjamingr commented:
@ronag talking to Matteo on FB: I think it's probably better to have a static `shedEventTarget` (name pending) method and probably not expose a getter/setter to avoid confusion about "what happens if we call a setter on multiple signals" and to avoid versioning issues.

benjamingr commented:
Naming suggestions welcome, also ccing @bterlson regarding naming suggestions or opinions here

benjamingr commented:
I've prototyped an alternative approach using a static method here: https://github.com/nodejs/node/pull/36061/commits/6a92d8c83221eaf4c96bb6649f77c5af242757fe - I think the DX is _slightly_ better though still not great?

mcollina commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,42 @@
+const lazyDOMException = hideStackFrames((message, name) => {
+  if (DOMException === undefined)
+    DOMException = internalBinding('messaging').DOMException;
+  return new DOMException(message, name);
```
This is going to be _very_ complicated for readable-stream. Could we avoid this?

benjamingr replied:
picked `AbortError` since [that's what the DOM suggests](https://dom.spec.whatwg.org/#abortcontroller-api-integration). However:

> Any web platform API using promises to represent operations that can be aborted must adhere to the following:

This is not a web platform API and we're not using promises here. So we really don't have to do this. 

---

That said: Why is this going to be complicated for ReadableStream? 

In browsers - this is quite easy and is just "new DOMException('The operation was aborted', 'AbortError')"

In Node - it's pretty easy to get an AbortError constructor from Node itself (for example by making an AbortError and then using .constructor)

mcollina replied:
I don't want this code using any `internal` stuff. This depends on `DomException` and `internalBinding`. Those should not be part of streams. I already have the Node error system done. Please use a normal error.

benjamingr replied:
I'll concede and change it to a regular error.

benjamingr commented:
@mcollina @ronag please take a look I've updated this PR based on the meeting consensus namely - streams are destroyed with an error

jasnell commented on doc/api/stream.md:
````diff
@@ -1799,6 +1799,32 @@ Calling `Readable.from(string)` or `Readable.from(buffer)` will not have
+);
+// Later, abort the operation closing the stream
+controller.abort();
+```
````
An example that shows this used with an async iterator would be good.

ronag replied:
I really don’t like this. Would much prefer a signal constructor argument.

jasnell commented on test/parallel/test-stream-readable-destroy.js:
```diff
@@ -268,3 +268,38 @@ const assert = require('assert');
+    for await (const chunk of read) {}
+  })(), /AbortError/);
+  setTimeout(() => controller.abort(), 0);
+}
```
A test that sets up a pipeline that uses AbortController would be good, as would a test using a single AbortController for two different Readables at the same time.

jasnell commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+    signal.addEventListener('abort', onAbort);
+    readable.on('close', () => signal.removeEventListener('abort', onAbort));
+  }
+  return readable;
```
Given that the listener is removed when the readable is closed, I'm not sure it's needed.

benjamingr replied:
No... why would we?

This method is explicitly "let the stream manage it" I think?

(If users really want to remove the listener they can with `getEventListeners` + `removeEventListener`)

mcollina replied:
Ok!

ronag commented on doc/api/stream.md:
```diff
@@ -1799,6 +1799,32 @@ Calling `Readable.from(string)` or `Readable.from(buffer)` will not have
 for performance reasons.
 
+### `stream.addAbortSignal(signal, readable)`
+<!-- YAML
```
Member
I really don’t like this. Would much prefer a signal constructor argument

benjamingr replied:
The problem is that not every API exposes construction easily or at all and doing this as a constructor argument means all the APIs would need to change

jasnell replied:
Yeah, it's not ideal but I've seen many streams subclasses that do not pass the constructor arguments down consistently. It may still be good to have a constructor argument in addition to this, but I think this api is unfortunately necessary.

mcollina replied:
Maybe we can have both?

benjamingr replied:
Happy to add both

benjamingr replied:
I'll keep this one `addAbortSignal` and follow up with another PR with the constructor argument if that's OK.

ronag replied:
If we keep this I think we need a way to remove the signal.

ronag replied:
continued in discussion below.

mcollina commented on test/parallel/test-stream-readable-destroy.js:
```diff
@@ -268,3 +268,38 @@ const assert = require('assert');
+  }));
+  controller.abort();
+  read.on('data', common.mustNotCall());
+}
```
Please add tests for writable as well.

benjamingr commented:
@mcollina I've changed the API to work with Readable and Writeable and not just Readable and added a test for that as well.

@jasnell added a test for `pipeline` and an example with async iterators in the README.

@ronag I'll add the signal as a constructor argument to `Readable` and `Writeable` like in the original iteration of this PR in a follow up PR.

PTAL :]

mcollina commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+module.exports = function addAbortSignal(signal, readable) {
```
this should not be called readable.

mcollina commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+  validateAbortSignal(signal, 'signal');
+  if (!isReadable(readable)) {
+    throw new ERR_INVALID_ARG_TYPE('readable', 'stream.Readable', readable);
+  }
```
It should be possible to abort a Writable.

mcollina commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+
+function isReadable(obj) {
+  return !!(obj && typeof obj.pipe === 'function');
+}
```
I don't think this works as you think it would. I think this passes for every descendant of Stream.

benjamingr replied:
I think so too and so do the tests - but I just copied `isReadable` from `pipeline`

mcollina commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+    onAbort();
+  } else {
+    signal.addEventListener('abort', onAbort);
+    readable.on('close', () => signal.removeEventListener('abort', onAbort));
```
This will not work for "legacy" streams.

ronag replied:
Yea, probably a good idea to use `stream.finished` here.

ronag commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+  return !!(obj && typeof obj.pipe === 'function');
+}
+
+module.exports = function addAbortSignal(signal, readable) {
```
I would prefer to have the args in the reverse order.

benjamingr replied:
Why?

ronag replied:
Usually the object that a function applies on/mutates is the first argument... which is how we usually do it. I think?

I guess as last argument I guess also makes sense in functional style with currying.

No strong opinion.

benjamingr replied:
Yeah this is so I can `addAbortSignal.bind(null, signal)` and run it over multiple streams but that's kind of a stretch and I have no real strong feelings either.

ronag commented on lib/internal/streams/readable.js:
```diff
@@ -50,11 +50,12 @@ const {
   ERR_STREAM_PUSH_AFTER_EOF,
   ERR_METHOD_NOT_IMPLEMENTED,
-  ERR_STREAM_UNSHIFT_AFTER_END_EVENT
+  ERR_STREAM_UNSHIFT_AFTER_END_EVENT,
```
Typo?

ronag commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,33 @@
+    signal.addEventListener('abort', onAbort);
+    readable.on('close', () => signal.removeEventListener('abort', onAbort));
+  }
+  return readable;
```
Return readable here seems unnecessary. Would be better to return a cleanup function?

benjamingr replied:
I think it really helps with ergonomics to return the stream here (see the tests for why). No very strong feelings though.

ronag replied:
That's fine but then we need a `removeAbortSignal` function.

benjamingr replied:
Why?

The case is:

 - I have a stream
 - I have a signal
 - I want the stream destroyed when I abort the controller linked to the signal

Why would I want to remove the listener from the stream? Is the use case for `removeAbortSignal` that the stream's lifespan is much longer than the signal's?

ronag replied:
Because you might run a function multiple times that does something and then changes it mind.


```js
function foo (stream) {
  const signal = ...
  stream.addAbortSignal(signal)
  readStreamUntilNextNewLine(stream)
  stream.removeAbortSignal(signal)
}

// Call foo in loop

```

benjamingr replied:
Why would you want to destroy the entire stream because you regret reading a single line? That sounds... odd no?

ronag replied:
If you can dynamically add an abort signal I would expect it to be possible to dynamically remove it.

benjamingr replied:
From a theoretical point of view I see it and from an implementation PoV this is easy to add but I just don't see any use case where people would use this.

I will happily concede with a single real use case

ronag replied:
> Why would you want to destroy the entire stream because you regret reading a single line? That sounds... odd no?

It could be a e.g. a timeout.

benjamingr replied:
But that's not what it would timeout - it would destroy the stream not cancel the request for data.

ronag replied:
Yes? 

BTW. I won't block because of this.

ronag replied:
I wouldn't implement it using this but I can see how e.g. it could be used for implementing a upstream retry mechanism where a remove function would be necessary to avoid memory leaks.

benjamingr replied:
As this is a month old already - I hope it's clear by now I am less concerned with landing this quickly and more with landing this right. I expected discussion (which is why this is deliberately kinda small)

Worth saying (_you_ know I'm on your side given I pushed for you in core in the first place but there might be more readers here) that I appreciate the feedback and discussion, there are maybe 10 people total in this discussion affecting APIs that impact millions of users.

> I wouldn't implement it using this but I can see how e.g. it could be used for implementing a upstream retry mechanism where a remove function would be necessary to avoid memory leaks.

A code example would be appreciated if you're up to make one. I honestly don't understand the use case. 

I tried (not for too long admittedly, like 20m) to come up with use cases for `removeAbortSignal` and couldn't find any that made sense.

I don't understand why I'd destroy the whole stream for "regretting" reading a line and in that case the feature request IIRC is passing a signal to "`read`" (in quotes since well, that wouldn't be possible/good given buffering, watermarks and read being sync and only consuming already ready data) and not to the stream itself.

benjamingr replied:
> there are maybe 10 people total in this discussion affecting APIs that impact millions of users.

That bit (in streams in particular) is always kind of terrifying ^^

ronag replied:
This is the best I can come up with at the moment:

```js
async function* read (signal, { start, end }) {
  const src = findInCacheOrCreateStream({ start }) // e.g. a cache of open range HTTP request
  src.addAbortSignal(signal)
  yield* readUntil(src, { end })
  if (!src.readableEnded) {
    // src can be used by other readers starting at 'end'.
    cacheStream(src, { start: end })
    src.removeAbortSignal(signal) 
  }
}
```

benjamingr replied:
It's a good example thanks.

I'm not sure I understand why I would want to destroy the entire cached stream because the user who called `read` decided to abort. That is, the thing we really want to abort here is the `readUntil` - and that would likely be need to still complete a single "read" rather than abort the read "in flight" so we know what `end` is in order to cache the stream.

I'll iterate a bit on the example and see if I can come up with a version where I want to destroy the whole stream (that is, I no longer want it to be in the cache for some reason?).

ronag replied:
> I'm not sure I understand why I would want to destroy the entire cached stream because the user who called read decided to abort.

Yea, good point.

I don't really have a better example at the moment.

ronag commented:
> @ronag I'll add the signal as a constructor argument to Readable and Writeable like in the original iteration of this PR in a follow up PR.

Why not in same PR? Pretty often these follow up PR's never happen... :/

benjamingr commented:
> Why not in same PR?

There is already a lot of discussion happening here and I don't feel like making _more_ changes here would make our lives easier basically.

benjamingr commented:
@mcollina @ronag addressed :]

ronag commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,34 @@
+    stream.destroy(new AbortError());
+  };
+  if (signal.aborted) {
+    onAbort();
```
@mcollina does this need to be in nextTick?

benjamingr replied:
Note cancellation is intentionally synchronous in this case to behave like .destroy(new AbortError)

Why would we want to nextTick here?

ronag replied:
I’m not sure. Just making sure we’ve considered it.

mcollina replied:
Not needed, events are nexticked already by destroy().

benjamingr commented:
What are my next steps here @nodejs/streams ?

mcollina commented on lib/internal/streams/add-abort-signal.js:
```diff
@@ -0,0 +1,34 @@
+
+const {
+  validateAbortSignal,
+} = require('internal/validators');
```
Can we avoid this? I prefer streams to be self-contained (readable-stream).

benjamingr replied:
We can inline the logic here this just does:

```js
const validateAbortSignal = hideStackFrames((signal, name) => {
  if (signal !== undefined &&
      (signal === null ||
       typeof signal !== 'object' ||
       !('aborted' in signal))) {
    throw new ERR_INVALID_ARG_TYPE(name, 'AbortSignal', signal);
  }
});
```

Would that be better?

mcollina replied:
Yes, thank you. Please add a comment on why this is inlined.

benjamingr replied:
Pushed, squashed, added a comment

mcollina commented on doc/api/stream.md:
```diff
@@ -1799,6 +1799,55 @@ Calling `Readable.from(string)` or `Readable.from(buffer)` will not have
+(async () => {
+  try {
+    for await (const chunk of stream) {
+      process(chunk);
```
```suggestion
      await process(chunk);
```

mcollina reviewed (approved):
lgtm

benjamingr commented:
@mcollina can you LGTM after the last rebase so I can land with the commit queue? I'm hesitant to land with ncu locally if I can help it :]

mcollina reviewed (approved):
lgtm

benjamingr commented:
Next up, the constructor arg version.
